{
  "$schema" : "https://json.schemastore.org/sarif-2.1.0.json",
  "version" : "2.1.0",
  "runs" : [ {
    "tool" : {
      "driver" : {
        "name" : "CodeQL",
        "organization" : "GitHub",
        "semanticVersion" : "2.17.0",
        "notifications" : [ {
          "id" : "py/baseline/expected-extracted-files",
          "name" : "py/baseline/expected-extracted-files",
          "shortDescription" : {
            "text" : "Expected extracted files"
          },
          "fullDescription" : {
            "text" : "Files appearing in the source archive that are expected to be extracted."
          },
          "defaultConfiguration" : {
            "enabled" : true
          },
          "properties" : {
            "tags" : [ "expected-extracted-files", "telemetry" ]
          }
        }, {
          "id" : "cli/database/analyze",
          "name" : "cli/database/analyze",
          "shortDescription" : {
            "text" : "CodeQL CLI: database analyze"
          },
          "fullDescription" : {
            "text" : "CodeQL CLI: database analyze"
          },
          "defaultConfiguration" : {
            "enabled" : true
          }
        } ],
        "rules" : [ {
          "id" : "py/mixed-returns",
          "name" : "py/mixed-returns",
          "shortDescription" : {
            "text" : "Explicit returns mixed with implicit (fall through) returns"
          },
          "fullDescription" : {
            "text" : "Mixing implicit and explicit returns indicates a likely error as implicit returns always return 'None'."
          },
          "defaultConfiguration" : {
            "enabled" : true,
            "level" : "note"
          },
          "properties" : {
            "tags" : [ "reliability", "maintainability" ],
            "description" : "Mixing implicit and explicit returns indicates a likely error as implicit returns always return 'None'.",
            "id" : "py/mixed-returns",
            "kind" : "problem",
            "name" : "Explicit returns mixed with implicit (fall through) returns",
            "precision" : "high",
            "problem.severity" : "recommendation",
            "sub-severity" : "high"
          }
        }, {
          "id" : "py/deprecated-slice-method",
          "name" : "py/deprecated-slice-method",
          "shortDescription" : {
            "text" : "Deprecated slice method"
          },
          "fullDescription" : {
            "text" : "Defining special methods for slicing has been deprecated since Python 2.0."
          },
          "defaultConfiguration" : {
            "enabled" : true,
            "level" : "warning"
          },
          "properties" : {
            "tags" : [ "maintainability" ],
            "description" : "Defining special methods for slicing has been deprecated since Python 2.0.",
            "id" : "py/deprecated-slice-method",
            "kind" : "problem",
            "name" : "Deprecated slice method",
            "precision" : "very-high",
            "problem.severity" : "warning",
            "sub-severity" : "high"
          }
        }, {
          "id" : "py/explicit-return-in-init",
          "name" : "py/explicit-return-in-init",
          "shortDescription" : {
            "text" : "`__init__` method returns a value"
          },
          "fullDescription" : {
            "text" : "Explicitly returning a value from an `__init__` method will raise a TypeError."
          },
          "defaultConfiguration" : {
            "enabled" : true,
            "level" : "error"
          },
          "properties" : {
            "tags" : [ "reliability", "correctness" ],
            "description" : "Explicitly returning a value from an `__init__` method will raise a TypeError.",
            "id" : "py/explicit-return-in-init",
            "kind" : "problem",
            "name" : "`__init__` method returns a value",
            "precision" : "very-high",
            "problem.severity" : "error",
            "sub-severity" : "low"
          }
        }, {
          "id" : "py/inheritance/incorrect-overriding-signature",
          "name" : "py/inheritance/incorrect-overriding-signature",
          "shortDescription" : {
            "text" : "Mismatch between signature and use of an overriding method"
          },
          "fullDescription" : {
            "text" : "Method has a different signature from the overridden method and, if it were called, would be likely to cause an error."
          },
          "defaultConfiguration" : {
            "enabled" : true,
            "level" : "error"
          },
          "properties" : {
            "tags" : [ "maintainability" ],
            "description" : "Method has a different signature from the overridden method and, if it were called, would be likely to cause an error.",
            "id" : "py/inheritance/incorrect-overriding-signature",
            "kind" : "problem",
            "name" : "Mismatch between signature and use of an overriding method",
            "precision" : "high",
            "problem.severity" : "error",
            "sub-severity" : "low"
          }
        }, {
          "id" : "py/inheritance/incorrect-overridden-signature",
          "name" : "py/inheritance/incorrect-overridden-signature",
          "shortDescription" : {
            "text" : "Mismatch between signature and use of an overridden method"
          },
          "fullDescription" : {
            "text" : "Method has a signature that differs from both the signature of its overriding methods and the arguments with which it is called, and if it were called, would be likely to cause an error."
          },
          "defaultConfiguration" : {
            "enabled" : true,
            "level" : "note"
          },
          "properties" : {
            "tags" : [ "maintainability" ],
            "description" : "Method has a signature that differs from both the signature of its overriding methods and\n              the arguments with which it is called, and if it were called, would be likely to cause an error.",
            "id" : "py/inheritance/incorrect-overridden-signature",
            "kind" : "problem",
            "name" : "Mismatch between signature and use of an overridden method",
            "precision" : "high",
            "problem.severity" : "recommendation",
            "sub-severity" : "high"
          }
        }, {
          "id" : "py/unexpected-raise-in-special-method",
          "name" : "py/unexpected-raise-in-special-method",
          "shortDescription" : {
            "text" : "Non-standard exception raised in special method"
          },
          "fullDescription" : {
            "text" : "Raising a non-standard exception in a special method alters the expected interface of that method."
          },
          "defaultConfiguration" : {
            "enabled" : true,
            "level" : "note"
          },
          "properties" : {
            "tags" : [ "reliability", "maintainability", "convention" ],
            "description" : "Raising a non-standard exception in a special method alters the expected interface of that method.",
            "id" : "py/unexpected-raise-in-special-method",
            "kind" : "problem",
            "name" : "Non-standard exception raised in special method",
            "precision" : "very-high",
            "problem.severity" : "recommendation",
            "sub-severity" : "high"
          }
        }, {
          "id" : "py/init-method-is-generator",
          "name" : "py/init-method-is-generator",
          "shortDescription" : {
            "text" : "`__init__` method is a generator"
          },
          "fullDescription" : {
            "text" : "`__init__` method is a generator."
          },
          "defaultConfiguration" : {
            "enabled" : true,
            "level" : "error"
          },
          "properties" : {
            "tags" : [ "reliability", "correctness" ],
            "description" : "`__init__` method is a generator.",
            "id" : "py/init-method-is-generator",
            "kind" : "problem",
            "name" : "`__init__` method is a generator",
            "precision" : "very-high",
            "problem.severity" : "error",
            "sub-severity" : "low"
          }
        }, {
          "id" : "py/iter-returns-non-iterator",
          "name" : "py/iter-returns-non-iterator",
          "shortDescription" : {
            "text" : "`__iter__` method returns a non-iterator"
          },
          "fullDescription" : {
            "text" : "The `__iter__` method returns a non-iterator which, if used in a 'for' loop, would raise a 'TypeError'."
          },
          "defaultConfiguration" : {
            "enabled" : true,
            "level" : "error"
          },
          "properties" : {
            "tags" : [ "reliability", "correctness" ],
            "description" : "The `__iter__` method returns a non-iterator which, if used in a 'for' loop, would raise a 'TypeError'.",
            "id" : "py/iter-returns-non-iterator",
            "kind" : "problem",
            "name" : "`__iter__` method returns a non-iterator",
            "precision" : "high",
            "problem.severity" : "error",
            "sub-severity" : "low"
          }
        }, {
          "id" : "py/iter-returns-non-self",
          "name" : "py/iter-returns-non-self",
          "shortDescription" : {
            "text" : "Iterator does not return self from `__iter__` method"
          },
          "fullDescription" : {
            "text" : "Iterator does not return self from `__iter__` method, violating the iterator protocol."
          },
          "defaultConfiguration" : {
            "enabled" : true,
            "level" : "error"
          },
          "properties" : {
            "tags" : [ "reliability", "correctness" ],
            "description" : "Iterator does not return self from `__iter__` method, violating the iterator protocol.",
            "id" : "py/iter-returns-non-self",
            "kind" : "problem",
            "name" : "Iterator does not return self from `__iter__` method",
            "precision" : "high",
            "problem.severity" : "error",
            "sub-severity" : "low"
          }
        }, {
          "id" : "py/modification-of-default-value",
          "name" : "py/modification-of-default-value",
          "shortDescription" : {
            "text" : "Modification of parameter with default"
          },
          "fullDescription" : {
            "text" : "Modifying the default value of a parameter can lead to unexpected results."
          },
          "defaultConfiguration" : {
            "enabled" : true,
            "level" : "error"
          },
          "properties" : {
            "tags" : [ "reliability", "maintainability" ],
            "description" : "Modifying the default value of a parameter can lead to unexpected\n              results.",
            "id" : "py/modification-of-default-value",
            "kind" : "path-problem",
            "name" : "Modification of parameter with default",
            "precision" : "high",
            "problem.severity" : "error",
            "sub-severity" : "low"
          }
        }, {
          "id" : "py/not-named-cls",
          "name" : "py/not-named-cls",
          "shortDescription" : {
            "text" : "First parameter of a class method is not named 'cls'"
          },
          "fullDescription" : {
            "text" : "Using an alternative name for the first parameter of a class method makes code more difficult to read; PEP8 states that the first parameter to class methods should be 'cls'."
          },
          "defaultConfiguration" : {
            "enabled" : true,
            "level" : "note"
          },
          "properties" : {
            "tags" : [ "maintainability", "readability", "convention" ],
            "description" : "Using an alternative name for the first parameter of a class method makes code more\n              difficult to read; PEP8 states that the first parameter to class methods should be 'cls'.",
            "id" : "py/not-named-cls",
            "kind" : "problem",
            "name" : "First parameter of a class method is not named 'cls'",
            "precision" : "high",
            "problem.severity" : "recommendation",
            "sub-severity" : "high"
          }
        }, {
          "id" : "py/not-named-self",
          "name" : "py/not-named-self",
          "shortDescription" : {
            "text" : "First parameter of a method is not named 'self'"
          },
          "fullDescription" : {
            "text" : "Using an alternative name for the first parameter of an instance method makes code more difficult to read; PEP8 states that the first parameter to instance methods should be 'self'."
          },
          "defaultConfiguration" : {
            "enabled" : true,
            "level" : "note"
          },
          "properties" : {
            "tags" : [ "maintainability", "readability", "convention" ],
            "description" : "Using an alternative name for the first parameter of an instance method makes\n              code more difficult to read; PEP8 states that the first parameter to instance\n              methods should be 'self'.",
            "id" : "py/not-named-self",
            "kind" : "problem",
            "name" : "First parameter of a method is not named 'self'",
            "precision" : "very-high",
            "problem.severity" : "recommendation",
            "sub-severity" : "high"
          }
        }, {
          "id" : "py/overly-complex-delete",
          "name" : "py/overly-complex-delete",
          "shortDescription" : {
            "text" : "Overly complex `__del__` method"
          },
          "fullDescription" : {
            "text" : "`__del__` methods may be called at arbitrary times, perhaps never called at all, and should be simple."
          },
          "defaultConfiguration" : {
            "enabled" : true,
            "level" : "note"
          },
          "properties" : {
            "tags" : [ "efficiency", "maintainability", "complexity", "statistical", "non-attributable" ],
            "description" : "`__del__` methods may be called at arbitrary times, perhaps never called at all, and should be simple.",
            "id" : "py/overly-complex-delete",
            "kind" : "problem",
            "name" : "Overly complex `__del__` method",
            "precision" : "high",
            "problem.severity" : "recommendation",
            "sub-severity" : "low"
          }
        }, {
          "id" : "py/mixed-tuple-returns",
          "name" : "py/mixed-tuple-returns",
          "shortDescription" : {
            "text" : "Returning tuples with varying lengths"
          },
          "fullDescription" : {
            "text" : "A function that potentially returns tuples of different lengths may indicate a problem."
          },
          "defaultConfiguration" : {
            "enabled" : true,
            "level" : "note"
          },
          "properties" : {
            "tags" : [ "reliability", "maintainability" ],
            "description" : "A function that potentially returns tuples of different lengths may indicate a problem.",
            "id" : "py/mixed-tuple-returns",
            "kind" : "problem",
            "name" : "Returning tuples with varying lengths",
            "precision" : "high",
            "problem.severity" : "recommendation",
            "sub-severity" : "high"
          }
        }, {
          "id" : "py/ignored-return-value",
          "name" : "py/ignored-return-value",
          "shortDescription" : {
            "text" : "Ignored return value"
          },
          "fullDescription" : {
            "text" : "Ignoring return values may result in discarding errors or loss of information."
          },
          "defaultConfiguration" : {
            "enabled" : true,
            "level" : "note"
          },
          "properties" : {
            "tags" : [ "reliability", "readability", "convention", "statistical", "non-attributable", "external/cwe/cwe-252" ],
            "description" : "Ignoring return values may result in discarding errors or loss of information.",
            "id" : "py/ignored-return-value",
            "kind" : "problem",
            "name" : "Ignored return value",
            "precision" : "medium",
            "problem.severity" : "recommendation",
            "sub-severity" : "high"
          }
        }, {
          "id" : "py/inheritance/signature-mismatch",
          "name" : "py/inheritance/signature-mismatch",
          "shortDescription" : {
            "text" : "Signature mismatch in overriding method"
          },
          "fullDescription" : {
            "text" : "Overriding a method without ensuring that both methods accept the same number and type of parameters has the potential to cause an error when there is a mismatch."
          },
          "defaultConfiguration" : {
            "enabled" : true,
            "level" : "warning"
          },
          "properties" : {
            "tags" : [ "reliability", "correctness" ],
            "description" : "Overriding a method without ensuring that both methods accept the same\n              number and type of parameters has the potential to cause an error when there is a mismatch.",
            "id" : "py/inheritance/signature-mismatch",
            "kind" : "problem",
            "name" : "Signature mismatch in overriding method",
            "precision" : "very-high",
            "problem.severity" : "warning",
            "sub-severity" : "high"
          }
        }, {
          "id" : "py/special-method-wrong-signature",
          "name" : "py/special-method-wrong-signature",
          "shortDescription" : {
            "text" : "Special method has incorrect signature"
          },
          "fullDescription" : {
            "text" : "Special method has incorrect signature"
          },
          "defaultConfiguration" : {
            "enabled" : true,
            "level" : "error"
          },
          "properties" : {
            "tags" : [ "reliability", "correctness" ],
            "description" : "Special method has incorrect signature",
            "id" : "py/special-method-wrong-signature",
            "kind" : "problem",
            "name" : "Special method has incorrect signature",
            "precision" : "high",
            "problem.severity" : "error",
            "sub-severity" : "low"
          }
        }, {
          "id" : "py/procedure-return-value-used",
          "name" : "py/procedure-return-value-used",
          "shortDescription" : {
            "text" : "Use of the return value of a procedure"
          },
          "fullDescription" : {
            "text" : "The return value of a procedure (a function that does not return a value) is used. This is confusing to the reader as the value (None) has no meaning."
          },
          "defaultConfiguration" : {
            "enabled" : true,
            "level" : "warning"
          },
          "properties" : {
            "tags" : [ "maintainability" ],
            "description" : "The return value of a procedure (a function that does not return a value) is used. This is confusing to the reader as the value (None) has no meaning.",
            "id" : "py/procedure-return-value-used",
            "kind" : "problem",
            "name" : "Use of the return value of a procedure",
            "precision" : "high",
            "problem.severity" : "warning",
            "sub-severity" : "low"
          }
        } ]
      },
      "extensions" : [ {
        "name" : "codeql/python-queries",
        "semanticVersion" : "0.9.13+839ca60f90f918b567e192642b0cc3003803f482",
        "locations" : [ {
          "uri" : "file:///C:/Users/minhd/.codeql/packages/codeql/python-queries/0.9.13/",
          "description" : {
            "text" : "The QL pack root directory."
          }
        }, {
          "uri" : "file:///C:/Users/minhd/.codeql/packages/codeql/python-queries/0.9.13/qlpack.yml",
          "description" : {
            "text" : "The QL pack definition file."
          }
        } ]
      } ]
    },
    "invocations" : [ {
      "toolExecutionNotifications" : [ {
        "locations" : [ {
          "physicalLocation" : {
            "artifactLocation" : {
              "uri" : "demo.py",
              "uriBaseId" : "%SRCROOT%",
              "index" : 0
            }
          }
        } ],
        "message" : {
          "text" : ""
        },
        "level" : "none",
        "descriptor" : {
          "id" : "py/baseline/expected-extracted-files",
          "index" : 0
        },
        "properties" : {
          "formattedMessage" : {
            "text" : ""
          }
        }
      }, {
        "message" : {
          "text" : "A fatal error occurred: Could not create access credentials for microsoft/coding-standards. Does this CodeQL pack exist on the public GitHub Container registry, or do you need to authenticate to access it? HTTP/1.1 403 Forbidden"
        },
        "level" : "error",
        "timeUtc" : "2024-04-18T18:58:40.861+00:00",
        "descriptor" : {
          "id" : "cli/database/analyze",
          "index" : 1
        },
        "properties" : {
          "attributes" : {
            "exitCode" : 2
          },
          "visibility" : {
            "telemetry" : false
          }
        }
      }, {
        "message" : {
          "text" : "A fatal error occurred: Could not create access credentials for microsoft/coding-standards. Does this CodeQL pack exist on the public GitHub Container registry, or do you need to authenticate to access it? HTTP/1.1 403 Forbidden"
        },
        "level" : "error",
        "timeUtc" : "2024-04-18T19:03:01.059+00:00",
        "descriptor" : {
          "id" : "cli/database/analyze",
          "index" : 1
        },
        "properties" : {
          "attributes" : {
            "exitCode" : 2
          },
          "visibility" : {
            "telemetry" : false
          }
        }
      }, {
        "message" : {
          "text" : "A fatal error occurred: Could not create access credentials for github/security-queries. Does this CodeQL pack exist on the public GitHub Container registry, or do you need to authenticate to access it? HTTP/1.1 403 Forbidden"
        },
        "level" : "error",
        "timeUtc" : "2024-04-18T19:05:07.982+00:00",
        "descriptor" : {
          "id" : "cli/database/analyze",
          "index" : 1
        },
        "properties" : {
          "attributes" : {
            "exitCode" : 2
          },
          "visibility" : {
            "telemetry" : false
          }
        }
      }, {
        "message" : {
          "text" : "A fatal error occurred: Could not write SARIF file\n(eventual cause: NoSuchFileException \"C:\\Users\\minhd\\python-analysis\\python-results.sarif\")"
        },
        "level" : "error",
        "timeUtc" : "2024-04-18T19:07:58.537+00:00",
        "descriptor" : {
          "id" : "cli/database/analyze",
          "index" : 1
        },
        "properties" : {
          "attributes" : {
            "exitCode" : 2
          },
          "visibility" : {
            "telemetry" : false
          }
        }
      } ],
      "executionSuccessful" : true
    } ],
    "artifacts" : [ {
      "location" : {
        "uri" : "demo.py",
        "uriBaseId" : "%SRCROOT%",
        "index" : 0
      }
    } ],
    "results" : [ ],
    "columnKind" : "unicodeCodePoints",
    "properties" : {
      "semmle.formatSpecifier" : "sarif-latest"
    }
  } ]
}